swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -558,4 +532,46 @@ extension Adapter: CallbackHandlerDelegate {
       }
     }
   }
+
+  public func getSystemDefaultResolvers() -> RustString {
+    var resolvers: [String] = Resolv().getservers().map(Resolv.getnameinfo)
+
+    // When the tunnel is up, we can only get the system's default resolvers
+    // by reading /etc/resolv.conf when matchDomains is set to a non-empty string.
+    // If matchDomains is an empty string, /etc/resolv.conf will contain connlib's
+    // sentinel, which isn't helpful to us.
+    // If networkSettings isn't initialized, this means we're setting up the tunnel,
+    // so we haven't set matchDomains at all yet, and so we
+    // don't need to do this dance and can just read what's in /etc/resolv.conf.
+    if let networkSettings = self.networkSettings {
+      // async / await can't be used here because this is an FFI callback
+      let semaphore = DispatchSemaphore(value: 0)
+      networkSettings.setMatchDomains(["firezone-fd0020211111"])
+
+      // Call apply with the completion handler
+      networkSettings.apply(
+        on: self.packetTunnelProvider, logger: self.logger,
+        completionHandler: { _ in
+          // Only now can we get the system resolvers
+          resolvers = Resolv().getservers().map(Resolv.getnameinfo)
+          semaphore.signal()
+        })
+
+      self.logger.log(
+        "Adapter.getSystemDefaultResolvers: Waiting on NetworkSettings.apply to complete")
+
+      semaphore.wait()
```

Comment:
I am not sure this is a good idea. You are halting all of connlib if you block in a callback.

Reply:
Yeah, the options here are either block in connlib's `get_system_default_resolvers`, or use another callback like @roop mentioned.

Blocking will be an issue if we ever call `get_system_default_resolvers` with packets in flight.

With the FFI refactoring planned in #3343, we can fix this like so:

- `Session.connect` passes initial dnsServers into the tunnel
- `Session.update(dnsServers: RustString, hasInternet: Boolean)` updates the tunnel with new dnsServers (optionally, connlib can choose to sleep if there's no internet -- for now this could just be ensuring the portal connection backoff timer doesn't grow). This is called by the app when switching networks.
- When connlib determines it needs new resolvers, for example if the upstream_dns were removed in the portal, it calls `get_system_default_resolvers`, which returns immediately but triggers the `Session.update` call above

`Session.update(dnsServers: [String], has_internet: Boolean)` call as the final callback in this dance, but right now we only call `get_system_default_resolvers` when setting up the tunnel and never again.

I think we can

Reply:
Actually, wait a second, does connlib ever need to call `get_system_default_resolvers`? We should be able to detect network changes on the app side and always send a `Session.update` when conditions change. Connlib then just stores these, and uses them if `upstream_dns` is not set from the portal.

Reply:
Worked through the details of the above proposal with @conectado and documented conclusions [here](https://github.com/firezone/firezone/issues/3343#issuecomment-1960057888).

As such, going to close this PR in favor of just solving this the right way (hopefully for the last time).

Reply:
btw, thank you @roop for the fixes! I'll make to incorporate them in my next PR.